merkledag/README.md

```diff
@@ -0,0 +1,144 @@
+
+The logical format -- in protobuf -- looks like this:
+
+```proto3
+// An IPFS MerkleDAG Link
+message MerkleLink {
+  bytes Hash = 1;   // multihash of the target object
+  string Name = 2;  // utf string name
+  uint64 Tsize = 3; // cumulative size of target object
+
+  // user extensions start at 50
+}
+
+// An IPFS MerkleDAG Node
+message MerkleNode {
+  repeated MerkleLink Links = 2; // refs to other objects
+  bytes Data = 1; // opaque user data
+
+  // user extensions start at 50
+}
+```
+
+### Serialized Format
+
+
+(TODO remove this? use only protobuf?)
+~~The logical representation is serialized into raw bytes using `multicodec`, a self-describing format that abstracts between serialization frameworks. That way, we can use the ipfs merkledag with various marshaling and serialization formats.~~
+
+The logical representation is serialized into raw bytes using _protocol buffers_, a serialization format.
+
+## Discussion
+
+### Real World Examples
+
+Many successful distributed systems employ specialized merkledags at their core:
+- merkle trees -- a special case of the merkle dag -- are a well known cryptographic technique used to authenticate large amounts of data. The original use case involved one-time lamport signatures.
+- SFS-RO turns a unix filesystem into a merkledag, constructing a secure, distributed filesystem.
+- git uses a merkledag to enable distributed version control and source code management. Other DVCSes, such as mercurial and monotone, also feature a merkledag.
+- plan9 uses a merkledag to construct its snapshotting filesystems -- Fossil and Venti.
+- bittorrent uses a merkledag to provide secure and short infohash links to its downloadable torrents.
+- Google Wave -- a distributed communications platform -- used a merkledag to construct its commutative operational transforms, and enable convergent distributed collaboration. This functionality has since been folded into Google Docs.
+- bitcoin uses a merkledag to construct the blockchain, a shared append-only ledger with convergent distributed consensus.
+- Tahoe-LAFS uses a merkledag to construct a secure, distributed, capability filesystem based on the least-authority principle.
+
+(NOTE: please suggest other systems to reference here.)
+
+### Thin Waist of Data Structures
+
+At its core, IPFS provides the merkledag as a primitive (or "internet layer") to build sophisticated applications easily. It is a "thin-waist" for secure, distributed applications, which -- by agreeing to follow the common format -- can then run across any replication, routing, and transport protocols. To draw an analogy, this is like the "thin-waist" IP provided to connect hosts across medium-specific networks.
+
+![](mdag.waist.png)
```

Comment:
IPRS is not the umbrella. IPRS is just a way to create records. it doesnt say anything about IPNS or making a name system. it merely specifies ways of writing + verifying records.

There's:
- IPNS - the name system for IPFS, which can resolve names on the ipfs routing system, dns names, and more.
- IPNS identity addresses (public key names mapped to records, stored on an IPFS routing system, as specified by this IPRS spec).

yes, DNS perhaps in the future should be separated out. this is a ton of work right now and not useful because it only confuses end users when they have to worry about multiple roots and so on. today, it's fine to use `"/ipns/..."` for `"/ipns/<hash-of-pubkey>/..."` or `"/ipns/<dns-domain-name/..."`. and really helps keep the number of logical symbols people have to worry about when they first encounter IPFS to a small number. 

note also that the word "IPFS" also means many things. it's not a big deal in practice.